File: src/initializer.js
const fs = require('fs').promises;
const path = require('path');
const natural = require('natural');
const ChatbotProcessor = require('./chatbotProcessor');

const stemmer = natural.PorterStemmer;
const tokenizer = new natural.WordTokenizer(); // For tokenizing before NGrams
// const stopwords = natural.stopwords; // Reverted

const MODEL_FILE = path.join(__dirname, 'models', 'logistic_regression_classifier.json');

let chatbotProcessorInstance = null; // Cache for the chatbot processor

async function initializeChatbot() {
    if (chatbotProcessorInstance) {
        console.log('Returning cached NLU classifier and chatbot processor.');
        return chatbotProcessorInstance;
    }

    let classifier;

    try {
        // Always try to load LogisticRegressionClassifier from file first
        try {
            await fs.access(MODEL_FILE);
            console.log('Loading NLU classifier (LogisticRegression) from file...');
            const data = await fs.readFile(MODEL_FILE, 'utf8');
            const modelData = JSON.parse(data);
            classifier = await new Promise((resolve, reject) => {
                natural.LogisticRegressionClassifier.restore(modelData, tokenizer.tokenize, stemmer.stem, (err, restoredClassifier) => {
                    if (err) reject(err); else resolve(restoredClassifier);
                });
            });
            console.log('NLU classifier (LogisticRegression) loaded.');
        } catch (err) {
            if (err.code !== 'ENOENT') throw err; // Re-throw if it's not a "file not found" error
            console.log('LogisticRegression model file not found or error during load, proceeding to train.');
        }

        if (!classifier) { // This block is entered if classifier was not loaded from file
            console.log('Training new NLU classifier (LogisticRegression with unigrams & bigrams & trigrams)...');
            const baseCorpus = JSON.parse(await fs.readFile(path.join(__dirname, 'data/training/base-corpus.json'), 'utf8'));
            const rhcpCorpus = JSON.parse(await fs.readFile(path.join(__dirname, 'data/training/rhcp-corpus.json'), 'utf8'));

            classifier = new natural.LogisticRegressionClassifier(stemmer);

            for (const corpus of [baseCorpus, rhcpCorpus]) {
                for (const item of corpus.data) {
                    if (item.intent !== 'None') {
                        for (const utterance of item.utterances) {
                            const tokens = tokenizer.tokenize(utterance.toLowerCase());
                            if (tokens && tokens.length > 0) {
                                let features = [];
                                features = features.concat(tokens); // Unigrams
                                if (tokens.length >= 2) { // Bigrams
                                    features = features.concat(natural.NGrams.bigrams(tokens).map(bi => bi.join('_')));
                                }
                                if (tokens.length >= 3) {
                                    features = features.concat(natural.NGrams.trigrams(tokens).map(tri => tri.join('_')));
                                }
                                if (features.length > 0) {
                                    classifier.addDocument(features, item.intent);
                                }
                            }
                        }
                    }
                }
            }
            console.time('classifier.train (LogisticRegression)');
            await classifier.train();
            console.timeEnd('classifier.train (LogisticRegression)');
            console.log('NLU classifier (LogisticRegression) trained.');

            // Always save the model if it was just trained (no longer conditional on NODE_ENV)
            const modelDataToSave = JSON.stringify(classifier);
            await fs.writeFile(MODEL_FILE, modelDataToSave, 'utf8');
            console.log('NLU classifier (LogisticRegression) saved to file.');
        }

        // Load static data regardless of training or loading model
        const bandInfo = JSON.parse(
            await fs.readFile(path.join(__dirname, 'data/static/band-info.json'), 'utf8')
        );
        const discography = JSON.parse(
            await fs.readFile(path.join(__dirname, 'data/static/discography.json'), 'utf8')
        );

        const trainingData = { // We still need training data for other parts of ChatbotProcessor
            base: JSON.parse(await fs.readFile(path.join(__dirname, 'data/training/base-corpus.json'), 'utf8')),
            rhcp: JSON.parse(await fs.readFile(path.join(__dirname, 'data/training/rhcp-corpus.json'), 'utf8'))
        };

        const staticData = {
            bandInfo,
            discography
        };

        const chatbotProcessor = new ChatbotProcessor(classifier, trainingData, staticData);
        chatbotProcessorInstance = chatbotProcessor; // Cache the instance

        console.log('Chatbot initialized successfully');
        return chatbotProcessorInstance;

    } catch (error) {
        console.error('Error initializing chatbot:', error);
        throw error;
    }
}

module.exports = { initializeChatbot }; 

File: rhcp-chatbot-py/app/chatbot/initializer.py
import json
import joblib
import os
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

from .processor import ChatbotProcessor

# --- File Paths ---
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
MODEL_FILE = os.path.join(BASE_DIR, '..', 'models', 'logistic_regression_classifier.joblib')
DATA_DIR = os.path.join(BASE_DIR, 'data')
STATIC_DATA_DIR = os.path.join(DATA_DIR, 'static')
TRAINING_DATA_DIR = os.path.join(DATA_DIR, 'training')

# --- Global Cache ---
chatbot_processor_instance = None

# --- Stemming and Tokenization ---
stemmer = PorterStemmer()

def stem_tokens(tokens):
    return [stemmer.stem(item) for item in tokens]

def tokenize(text):
    return stem_tokens(word_tokenize(text.lower()))


def load_json_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def train_classifier(training_files):
    print("Training new NLU classifier (LogisticRegression with Tfidf)...")
    
    texts = []
    intents = []

    for file_path in training_files:
        corpus = load_json_file(file_path)
        for item in corpus['data']:
            if item['intent'] != 'None':
                for utterance in item['utterances']:
                    texts.append(utterance)
                    intents.append(item['intent'])

    # Create a scikit-learn pipeline
    # TfidfVectorizer will handle tokenization, n-grams, and TF-IDF weighting
    # LogisticRegression is the classifier
    pipeline = Pipeline([
        ('tfidf', TfidfVectorizer(tokenizer=tokenize, ngram_range=(1, 3))),
        ('clf', LogisticRegression(random_state=42, solver='lbfgs', multi_class='auto'))
    ])

    print("Training the pipeline...")
    pipeline.fit(texts, intents)
    print("Training complete.")

    # Save the trained pipeline
    joblib.dump(pipeline, MODEL_FILE)
    print(f"Classifier saved to {MODEL_FILE}")

    return pipeline

async def initialize_chatbot():
    global chatbot_processor_instance
    if chatbot_processor_instance:
        print("Returning cached chatbot processor.")
        return chatbot_processor_instance

    classifier = None
    
    # 1. Try to load the classifier
    if os.path.exists(MODEL_FILE):
        print(f"Loading NLU classifier from {MODEL_FILE}...")
        try:
            classifier = joblib.load(MODEL_FILE)
            print("NLU classifier loaded successfully.")
        except Exception as e:
            print(f"Error loading model: {e}. Retraining...")
            classifier = None

    # 2. If loading failed, train a new one
    if not classifier:
        training_files = [
            os.path.join(TRAINING_DATA_DIR, 'base-corpus.json'),
            os.path.join(TRAINING_DATA_DIR, 'rhcp-corpus.json')
        ]
        classifier = train_classifier(training_files)

    # 3. Load static and training data for the processor
    band_info = load_json_file(os.path.join(STATIC_DATA_DIR, 'band-info.json'))
    discography = load_json_file(os.path.join(STATIC_DATA_DIR, 'discography.json'))
    
    base_corpus = load_json_file(os.path.join(TRAINING_DATA_DIR, 'base-corpus.json'))
    rhcp_corpus = load_json_file(os.path.join(TRAINING_DATA_DIR, 'rhcp-corpus.json'))

    training_data = {'base': base_corpus, 'rhcp': rhcp_corpus}
    static_data = {'bandInfo': band_info, 'discography': discography}

    # 4. Create the ChatbotProcessor instance
    chatbot_processor_instance = ChatbotProcessor(classifier, training_data, static_data)
    print("Chatbot initialized successfully.")
    
    return chatbot_processor_instance 